### PR TITLE
fix: SAML fixes for uppercase email & GOOGLE → SAML idp switch

### LINKED ISSUE #14631: [CAL-3485] SAML/SSO auto-linking failing for GOOGLE auth users
When trying to login via SAML/SSO when the idP is set to google, it fails to login and link the user to SAML, and results in error.

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-3485](https://linear.app/calcom/issue/CAL-3485/samlsso-auto-linking-failing-for-google-auth-users)</sub>

### PR DESCRIPTION
**Pull request by alishaz-polymath**
## What does this PR do?

- Fixes the bug where if the idP (SAML) had uppercase email, when switching from other idP to SAML on Cal, we'd use the uppercase email from the idP and thus mess up the email matchup, sometimes even creating a duplicate account.
- Adds the feature now allowing users to switch from GOOGLE idP to SAML for existing users

- Fixes #14631 #14833 
- Fixes CAL-3485
- Fixes CAL-3600

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected)
- [x] I have added a Docs issue [here](https://github.com/calcom/docs/issues/62) if this PR makes changes that would require a [documentation change](https://docs.cal.com)
- [ ] I have added or modified automated tests that prove my fix is effective or that my feature works (PRs might be rejected if logical changes are not properly tested)

## How should this be tested?

- Are there environment variables that should be set?
- What are the minimal test data to have?
- What is expected (happy path) to have (input and output)?
- Any other important info that could help to test that PR

## Checklist

- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my changes generate no new warnings

### CODE REVIEW

File: packages/features/auth/lib/next-auth-options.ts [hunk 1]
```diff
@@ -342,7 +342,8 @@ if (isSAMLLoginEnabled) {
           return null;
         }
 
-        const { id, firstName, lastName, email } = userInfo;
+        const { id, firstName, lastName } = userInfo;
+        const email = userInfo.email.toLowerCase();
```
Comment (alishaz-polymath) [inline packages/features/auth/lib/next-auth-options.ts:346]: We convert to lower case to ensure an uppercase email doesn't slip in from the idP

File: packages/features/auth/lib/next-auth-options.ts [hunk 2]
```diff
@@ -844,7 +845,7 @@ export const AUTH_OPTIONS: AuthOptions = {
               where: { email: existingUserWithEmail.email },
               // also update email to the IdP email
               data: {
-                email: user.email,
+                email: user.email.toLowerCase(),
```
Comment (alishaz-polymath) [inline packages/features/auth/lib/next-auth-options.ts:848]: We convert to lower case to ensure an uppercase email doesn't slip in from the idP into our DB

File: packages/features/auth/lib/next-auth-options.ts [hunk 3]
```diff
@@ -857,6 +858,19 @@ export const AUTH_OPTIONS: AuthOptions = {
...
             return "/auth/error?error=use-password-login";
+          } else if (
+            existingUserWithEmail.identityProvider === IdentityProvider.GOOGLE &&
+            idP === IdentityProvider.SAML
+          ) {
+            await prisma.user.update({
+              where: { email: existingUserWithEmail.email },
+              // also update email to the IdP email
+              data: {
+                email: user.email.toLowerCase(),
```
Comment (alishaz-polymath) [inline packages/features/auth/lib/next-auth-options.ts:869]: We convert to lower case to ensure an uppercase email doesn't slip in from the idP into our DB

File: packages/features/auth/lib/next-auth-options.ts [hunk 4]
```diff
@@ -857,6 +858,19 @@ export const AUTH_OPTIONS: AuthOptions = {
...
+                identityProvider: idP,
+                identityProviderId: account.providerAccountId,
+              },
```
Comment (alishaz-polymath) [inline packages/features/auth/lib/next-auth-options.ts:872]: Added the check to allow switching from GOOGLE idP to SAML, which was earlier throwing use-identity-login error.

**Review by emrysal — APPROVED** [review]
Looks good 👍

**Review by pumfleet — APPROVED** [review]
Code looks good to me

**Pull request merged**